src/approaches/nsrt_rl_approach.py

```diff
@@ -72,10 +90,128 @@ def learn_from_interaction_results(
+            curr_states = [traj.states[0]]
+            curr_actions = []
+            curr_rewards = []
+            curr_relative_params = []
+            actions = (a for a in traj.actions)
+            block = [b for b in curr_option.objects if b.type.name=='block'][0]
+            robot = [r for r in curr_option.objects if r.type.name=='robot'][0]
+
+            # Generate transition data, (s, a, s', r, relative_param). The
+            # reward R(s, a, s') = neg_reward if s' is not within epsilon of the
+            # subgoal, and pos_reward otherwise.
+            for j, s in enumerate(traj.states[1:]):
+                curr_states.append(s)
+                curr_actions.append(next(actions))
+                # TODO: is there a better way to do this based on
+                # curr_option.parent._changing_obj_feats? Couldn't access this
+                # field, but could access curr_option.parent._changing_var_order
+                if curr_option.params[-1] > 0:  # if holding becomes true (pick)
+                    changing_obj_feats = np.array([
+                        s.get(block, 'grasp'),
+                        s.get(robot, 'x'),
+                        s.get(robot, 'y'),
+                        s.get(robot, 'grip'),
+                        s.get(robot, 'holding')
+                    ])
+                else:
+                    changing_obj_feats = np.array([
+                        s.get(block, 'x'),
+                        s.get(block, 'grasp'),
+                        s.get(robot, 'x'),
+                        s.get(robot, 'grip'),
+                        s.get(robot, 'holding')
+                    ])
+                curr_relative_params.append(curr_option.memory["absolute_params"] - changing_obj_feats)
+
+                if curr_option.terminal(s):
+                    # TODO: make this not hardcoded to be environment specific,
+                    # or implement specifically per environment and throw an
+                    # error for environments where this is not implemented.
+
+                    # Check if we reached our subgoal within a tolerance by
+                    # checking the difference between our proposed subgoal
+                    # (which is expressed relatively) and the relative changes
+                    # that actually happened.
+                    if curr_option.params[-1] > 0:  # if holding becomes true
+                        dblock = self.infer_delta(block, curr_states, ['grasp'])
+                        drobot = self.infer_delta(robot, curr_states, ['x', 'y', 'grip', 'holding'])
+                    else:
+                        dblock = self.infer_delta(block, curr_states, ['x', 'grasp'])
+                        drobot = self.infer_delta(robot, curr_states, ['x', 'grip', 'holding'])
+                    actual_delta = np.array(dblock + drobot)
+                    if np.allclose(curr_option.params, actual_delta, atol=self._reward_epsilon):
```

Comment:
another thing we will probably want to do is give a large negative reward if any objects that are not supposed to change during this option execution do actually change. we can leave a TODO in the code or just remember

Reply:
Added a TODO as a comment. This will also be implemented easily after we figure out the thing related to `self._changing_var_to_feat`.